python/lsst/obs/base/ingest.py

```diff
@@ -355,16 +419,202 @@ def _calculate_dataset_info(self, header, filename):
+            directory, file_in_dir = os.path.split(path)
+            files_by_directory[directory].add(file_in_dir)
+
+        # All the metadata read from index files with keys of full path
+        index_entries = {}
+
+        # Index files we failed to read
+        bad_index_files = set()
+
+        # Any good index files that were found and used
+        good_index_files = set()
+
+        # Look for index files in those directories
+        for directory, files_in_directory in files_by_directory.items():
+            possible_index_file = os.path.join(directory, index_root_file)
+            if os.path.exists(possible_index_file):
+                # If we are explicitly requesting an index file the
+                # messages should be different
+                index_msg = "inferred"
+                is_implied = True
+                if index_root_file in files_in_directory:
+                    index_msg = "explicit"
+                    is_implied = False
+
+                # Try to read the index file and catch and report any
+                # problems.
+                try:
+                    index = read_index(possible_index_file, force_dict=True)
+                except Exception as e:
+                    # For now only trigger the callback if the index file
+                    # was asked for explicitly.
+                    if not is_implied:
+                        self._on_metadata_failure(possible_index_file, e)
+                    if self.config.failFast:
+                        raise RuntimeError(f"Problem reading index file from {index_msg} "
+                                           f"location {possible_index_file}") from e
+                    bad_index_files.add(possible_index_file)
+                    continue
+
+                self.log.debug("Extracted index metadata from %s file %s", index_msg, possible_index_file)
+                good_index_files.add(possible_index_file)
+
+                # Go through the index adding entries for files.
+                # If we have non-index files in this directory marked for
+                # ingest we should only get index information for those.
+                # If the index file was explicit we use all entries.
+                if is_implied:
+                    files_to_ingest = files_in_directory
+                else:
+                    files_to_ingest = set(index)
+
+                # Copy relevant metadata into a single dict for all index
+                # entries.
+                for file_in_dir in files_to_ingest:
+                    # Skip an explicitly specified index file.
+                    # This should never happen because an explicit index
+                    # file will force ingest of all files in the index
+                    # and not use the explicit file list. If somehow
+                    # this is not true we continue. Raising an exception
+                    # seems like the wrong thing to do since this is harmless.
```

Comment:
"harmless" could mean "someone specified a bad file wildcard to make their list and maybe they're getting more things than they bargained for? Worth at least a `log.debug`, if not a `log.info`?